sympy/codegen/abstract_nodes.py

```diff
@@ -0,0 +1,15 @@
+"""This module provides containers for python objects that are valid
+printing targets but are not a subclass of SymPy's Printable.
+"""
+
+
+from sympy.core.containers import Tuple
+
+
+class List(Tuple):
+    """Represents a (frozen) (Python) list (for code printing purposes)."""
+    def __eq__(self, other):
```

Comment:
Is the `__eq__` method needed for this class to function or is it just for the tests?

In general it's better if Basic subclass don't override `__eq__`.

Reply:
```python
from sympy import *
class List(Tuple):
    pass
print([1] == List(1))

class List(Tuple):
    """Represents a (frozen) (Python) list (for code printing purposes)."""
    def __eq__(self, other):
        if isinstance(other, list):
            return self == List(*other)
        else:
            return self.args == other
print([1] == List(1))
```
First case is `False`, second is `True`, so I do believe that is needed here.

Reply:
```python
from sympy import *
class List(Tuple):
    pass
print([1] == List(1))

class List(Tuple):
    """Represents a (frozen) (Python) list (for code printing purposes)."""
    def __eq__(self, other):
        if isinstance(other, list):
            return self == List(*other)
        else:
            return self.args == other
print([1] == List(1))
```
Shows that the first case is `False` while the second case is `True`

Reply:
Yes, but why does that matter? Does the List class actually need to be compared (with `==`) to the list class in order to function properly in actual use?

Reply:
I'm not sure, that would have to be tested. Note that I only moved this class. It already had this functionality, so I assume there was a reason (perhaps a bit naive). It was previously in pynodes.py, which should indicate that this is specific to python printers, which it actually is not since it was already being used for all `CodePrinter`s. The 'bug' related to this class was actually that it had to be used earlier in `CodePrinter.doprint` to avoid converting `Tuple`s with `list`s as arguments.

Reply:
> Note that I only moved this class

Oh, okay. Let's just leave it for now then.

Reply:
I think we can remove the __eq__ method and the test if we want. I didn't reflect over implications of overriding __eq__ in a Basic subclass.